### PR TITLE
Wait for confirmations, fixes + refactoring

### LINKED ISSUE #261: Wait for confirmations before accounting
Currently there is no waiting for confirmation period, so there is big chance your tx has been mined in the uncle block.

Both claiming and execute should wait for 12 confirmations before accounting and announcing the successful operation.

### PR DESCRIPTION
**Pull request by kosecki123**
* Waits for confirmations (6) before accounting and considering transaction OK
* Refactor code for proper unit tests (removes ganache dependency) 
* Introduced snapshots for integration tests to make them repeatable 
* Fixes potential race conditions by introducing cache scanner debouncing.
* SonarTS static analysis

Fixes #261

### CODE REVIEW

File: src/Actions/Pending.ts [hunk 1]
```diff
@@ -1,100 +1,106 @@
...
-const hasPending = async (
-  conf: Config,
-  txRequest: ITxRequestPending,
-  opts: PendingOpts
-): Promise<boolean> => {
-  let result: boolean = false;
-  if (conf.txPool && conf.txPool.running()) {
-    result = await hasPendingPool(conf, txRequest, opts);
+  /**
+   * Uses the Geth specific RPC request `txpool_content` to search
```
Comment (lsaether) [inline src/Actions/Pending.ts]: Comment doesn't match here.

File: src/Actions/Pending.ts [hunk 2]
```diff
@@ -1,100 +1,106 @@
...
+    let result: boolean = false;
+    if (this.txPool.running()) {
+      result = await this.hasPendingPool(txRequest, opts);
+    }
+    return result;
   }
 
-  return true; //if there is an error, assume tq exists so we don't loose
-};
+  /**
```
Comment (lsaether) [inline src/Actions/Pending.ts]: Comment doesn't match.

File: src/Actions/Pending.ts [hunk 3]
```diff
@@ -1,100 +1,106 @@
...
-    const currentGasPrice: BigNumber = await conf.util.networkGasPrice();
-    validPending = conf.txPool.pool.get(txRequest.address, 'to').filter((tx: ITxPoolTxDetails) => {
-      const withValidGasPrice =
-        !opts.checkGasPrice || hasValidGasPrice(currentGasPrice, tx, opts.minPrice);
-      return isOfType(tx, opts.type) && withValidGasPrice;
-    });
-    return validPending.length > 0;
-  } catch (e) {
-    conf.logger.info(e);
+  /**
```
Comment (lsaether) [inline src/Actions/Pending.ts]: Comment doesn't match.

File: src/Actions/Pending.ts [hunk 4]
```diff
@@ -1,100 +1,106 @@
...
+    }
+    return transaction.input === FnSignatures[type];
   }
-  return result;
-};
 
-export default hasPending;
+  /**
+   * Checks that pending transactions in the transaction pool have valid gasPrices.
+   * @param {Config} conf Config object.
```
Comment (lsaether) [inline src/Actions/Pending.ts]: Comment is not updated.

File: src/Actions/Pending.ts (hunk 3 above)
Comment (kosecki123) [reply src/Actions/Pending.ts]: Fixed

File: test/unit/UnitTestActions.ts [hunk 5]
```diff
@@ -11,50 +7,46 @@ import {
...
-});
+//     const nextAccount = config.wallet.nextAccount.getAddressString();
+//     const claimingResult = await actions.claim(tx, nextAccount);
+
+//     assert.equal(timenode.getClaimedNotExecutedTransactions()[myAccount].length, 0);
+//     assert.equal(timenode.getUnsucessfullyClaimedTransactions()[myAccount].length, 1);
+
+//     assert.equal(claimingResult, ClaimStatus.FAILED);
+//   });
+// });
```
Comment (Bagaric) [inline test/unit/UnitTestActions.ts]: What's the issue with these tests?

File: test/unit/UnitTestEconomicStrategy.ts [hunk 6]
```diff
@@ -1,119 +1,246 @@
...
 
 describe('Economic Strategy Tests', () => {
-  let config: Config;
-  let txTimestamp: any;
+  let economicStrategyManager: IEconomicStrategyManager;
+
+  const account = '0x123456';
+  const defaultBalance = new BigNumber(10000000);
+  const defaultBounty = new BigNumber(10000000);
+  const defaultGasPrice = new BigNumber(10000000);
```
Comment (Bagaric) [inline test/unit/UnitTestEconomicStrategy.ts]: Might be better to make this a single variable `const tenMwei = new BigNumber(10000000);`.

**Review by josipbagaric — APPROVED** [review]
Looks good - minor comments.

File: test/unit/UnitTestEconomicStrategy.ts [hunk 7]
```diff
@@ -12,13 +12,15 @@ import {
 import { EconomicStrategyStatus } from '../../src/Enum';
 import { ITxRequest } from '../../src/Types';
 
+// tslint:disable-next-line:no-big-function
 describe('Economic Strategy Tests', () => {
   let economicStrategyManager: IEconomicStrategyManager;
+  const MWei = new BigNumber(10000000);
```
Comment (Bagaric) [inline test/unit/UnitTestEconomicStrategy.ts]: This is 10 mwei, isn't it? The name `MWei` might be misleading.

Comment (kosecki123) [reply test/unit/UnitTestEconomicStrategy.ts]: Ow. That's true!

**Pull request merged**